addon/search/search.js

```diff
@@ -76,19 +116,18 @@
     var state = getSearchState(cm);
     if (state.query) return findNext(cm, rev);
     var q = cm.getSelection() || state.lastQuery;
-    dialog(cm, queryDialog, "Search for:", q, function(query) {
+    dialog(cm, queryDialog, "Search for:", q, function(queryStr, keyEvent) {
       cm.operation(function() {
-        if (!query || state.query) return;
-        state.query = parseQuery(query);
-        cm.removeOverlay(state.overlay, queryCaseInsensitive(state.query));
-        state.overlay = searchOverlay(state.query, queryCaseInsensitive(state.query));
-        cm.addOverlay(state.overlay);
-        if (cm.showMatchesOnScrollbar) {
-          if (state.annotate) { state.annotate.clear(); state.annotate = null; }
-          state.annotate = cm.showMatchesOnScrollbar(state.query, queryCaseInsensitive(state.query));
+        initQuery(cm, queryStr);
+
+        if (cm.state.incrementalSearch) {
+          // In incremental search we need to prevent default behavior
+          // from pre-maturely closing the dialog box.
+          CodeMirror.e_stop(keyEvent);
+          findNext(cm, (keyEvent.shiftKey === true));
```

Comment:
That's not quite right.

So in both the original and this PR version, the dialog executes the search callback once it detects an Enter keydown event.  This is where they start to diverge:

In original:
The search callback executes and parses the query into SearchState.  Then the global keybindings, i.e. Cmd-G / Shift-Cmd-G are bound to the search commands to move next / prev.

In PR:
The search callback executes and will only parse the query into SearchState if the query string has changed from the previous execution.  This was necessary because the dialog remains on screen, and the user may/may not choose to modify the search.  Because the dialog is still on screen, the dialog's local keybindings are still active and every keydown of Enter will trigger the search callback which will move to next Search term.

Both versions are pretty much the same thing.  The new one just uses Enter/Shift-Enter to search next/previous and it also keeps dialog visible.

About packaging this into a command...  How would you see this command used?  I see the search's findNext / findPrev commands being bound to the global keybindings Cmd-G / Shift-Cmd-G.  Because Enter is an already bound key in the global context, I don't see how we could leverage it for search.  It works today, because the Dialog provides it's own keybindings that override those defined at global level.